clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -476,7 +476,6 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 else if (!future.isRetriable())
                     throw exception;
 
-                resetStateAndRejoin(String.format("rebalance failed with retriable error %s", exception));
```

Comment:
My reasoning here is the following. First, there is no reason to reset the memberId/generation if it is a retriable error. For most transient errors, our session timeout will not have expired, so it is better to rejoin with the same memberId. This ensures that we will not have an extra delay while the old memberId gets expired. On the other hand, if the session timeout _did_ expire, then our next `JoinGroup` will fail with `UNKNOWN_MEMBER_ID`, which will cause our state to be reset in the `JoinGroupResponseHandler`.

Second, there should be no need to reset the `rejoinNeeded` flag. Once we begin a rebalance, this will remain set to `true` until the `SyncGroup` completes successfully which consequently causes the future to be completed successfully. Hence there's no way that I can see for `rejoinNeeded` to be `false` if the future has failed.

Reply:
Your reasoning makes sense to me. From a first read, the PR looks pretty good. I will make a second pass on Monday to ensure that I cover all the cases.

Reply:
Just adding my 2c here: The reason I add that resetting in https://github.com/apache/kafka/commit/7e7bb184d2abe34280a7f0eb0f0d9fc0e32389f2# is that when rebalance failed with retriable exceptions, we would not trigger the `onJoinPrepare(generation.generationId, generation.memberId)` again upon retry. However for certain cases we do want to trigger that function in order to e.g. revoke partitions.

Reading that change again, I agree this is probably not the best pattern anyways, probably we can consider resetting `needsJoinPrepare` in the `!future.succeeded()` case as well before we retry, and then remove this line?

Reply:
@guozhangwang Thanks. Could you mention a case of a retriable exception where we would need to call `onJoinPrepare`?  At a high level, I think we should only need `onJoinPrepare` if there has been some change to the membership. If we fall out of the group after an `UNKNOWN_MEMBER_ID` error, then we need `onJoinPrepare`, but if we're retrying after a disconnect, I can't think why we need it.

Reply:
Discussed offline. The case we're trying to handle comes from https://issues.apache.org/jira/browse/KAFKA-12983. I am going to look into resetting the `needsJoinPrepare` flag in `requestRejoin` as an alternative to calling `resetStateAndRejoin` here.

Reply:
@guozhangwang @ableegoldman After looking into this issue, my conclusion is that no additional changes are needed beyond the one line removal here.  If the subscription/metadata changes while the rebalance is in progress, then we will detect it in `onJoinComplete` and we will call `requestRejoin`. But after `onJoinComplete` returns, we already reset `needsJoinPrepare`, so we should not miss the call to `onJoinPrepare` after we send the next `JoinGroup`. 

On the other hand, if the rebalance fails because of an error such as UNKNOWN_MEMBER_ID, then `JoinResponseHandler` already invokes `resetGenerationOnResponseError`. After the fix from KAFKA-12983, this ensures that `needsJoinPrepare` gets reset.

Are there any other cases to consider?

Reply:
After reading https://github.com/apache/kafka/pull/10986 again I think I agree with you. The original corner case is that during a rebalance if the hb thread detect we've already lost the membership, i.e. falling in line 458, we should also reset the flag. We would not need to do more than that.

About removing this line as a whole, one nit thought we probably still want to reset `rejoinNeeded` here --- this is probably not a correctness issue since in `rejoinNeededOrPending` we check both conditions `rejoinNeeded || joinFuture != null;` and the second is still true. But I feel resetting it makes the logic a bit cleaner.

Reply:
I have double checked the error handling as well and I do agree with the change as all the errors, which requires to rejoin, are explicitly handled in the `JoinResponseHandler` and the `SyncGroupResponseHandler`. The change looks good to me.

I do agree with @guozhangwang that reseting `rejoinNeeded` would make the whole logic a bit cleaner.

Reply:
Thanks for the suggestion. Updated.